arch/x86/deflate_quick.c

```diff
@@ -237,7 +243,7 @@ ZLIB_INTERNAL block_state deflate_quick(deflate_state *s, int flush) {
             hash_head = quick_insert_string(s, s->strstart);
             dist = s->strstart - hash_head;
 
-            if (dist > 0 && (dist-1) < (s->w_size - 1)) {
+            if (dist > 0 && (dist-1) < wsize) {
```

Comment:
Should be ```wsize - 1```...

Reply:
I think here it should be `wsize`. I don't see the need for `-1` .  It is already a less than check.

Reply:
Obviously both matches need to fit in the window at the same time, otherwise the hash could not be calculated. The test is incorrect, but removing ``` - 1``` makes it less correct... Obviously ```dist``` can't be more than ```wsize - MIN_MATCH```.

Reply:
But then why wasn't the original check `s->w_size - 3`?

Reply:
> In the worst case the original string starts at byte index 0, and match can be last 3 bytes of window (8189, 8190, 8191). I don't know why it isn't w_size - 3, that is why I said the original test is also incorrect...

I think we should only care that `dist-1` is less than 8192  or `quick_dist_codes` array size, because in `static_emit_ptr`:
```
    unsigned code2 = quick_dist_codes[dist-1] >> 8;
    unsigned len2  = quick_dist_codes[dist-1] & 0xFF;
```

Reply:
I think we should add a comment here why it uses ```dist - 1``` in the second test... That way people don't need to scan the whole source to see the reason...

Reply:
I think I found the reason which is that `hash_mask` is window size -1.

Reply:
I have kept the -1.